main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -541,16 +541,16 @@ object Resolver {
     * Performs name resolution on the given enum `e0` in the given namespace `ns0`.
     */
   def resolveEnum(e0: NamedAst.Declaration.Enum, env0: ListMap[String, Resolution], taenv: Map[Symbol.TypeAliasSym, ResolvedAst.Declaration.TypeAlias], ns0: Name.NName, root: NamedAst.Root)(implicit flix: Flix): Validation[ResolvedAst.Declaration.Enum, ResolutionError] = e0 match {
-    case NamedAst.Declaration.Enum(doc, ann, mod, sym, tparams0, derives0, cases0, loc) =>
+    case NamedAst.Declaration.Enum(doc, ann, mod, sym, tparams0, NamedAst.Derivations(classes, loc1), cases0, loc2) =>
       val tparamsVal = resolveTypeParams(tparams0, env0, ns0, root)
       flatMapN(tparamsVal) {
         case tparams =>
           val env = env0 ++ mkTypeParamEnv(tparams.tparams)
-          val derivesVal = resolveDerivations(derives0, env, ns0, root)
+          val derivesVal = resolveDerivations(classes, env, ns0, root)
```

Comment:
Since we already have a dedicated function, can we just take the derives as input?

Reply:
I think its because we don't like the nested pattern match.

Instead just change the signature of `resolveDerivations` then it takes a `derivations` value. That seems reasonable, no?

Reply:
We could, but again, then we'd just have to pull out both of the fields at the end

Reply:
I don't understand. The point is to avoid a deep pattern match and have resolveDerivations take the derivation type.

Reply:
This way it is structurally recursive: it means that the control-flow follows the structure of the data.

Reply:
Also, lets say we extend Derivations with more information then that would immediately be available inside resolveDerivations

Reply:
I just mean that we can either do
```
case NamedAst.Declaration.Enum(... NamedAst.Derivations(classes, dloc), ...) =>
  ...
  resolveDerivations(classes, ...)
  ...
  ResolvedAst.Declaration.Enum(... Ast.Derivations(derives, dloc), ...)
```
or
```
case NamedAst.Declaration.Enum(... derives0, ...) =>
  ...
  resolveDerivations(derives0, ...)
  ...
  ResolvedAst.Declaration.Enum(... Ast.Derivations(derives, derives0.loc), ...)
```

We can certainly do it like that, but I'm just trying to say that no matter what it's not just going to be
```
case NamedAst.Declaration.Enum(... derives0, ...) =>
  ...
  resolveDerivations(derives0, ...)
  ...
  ResolvedAst.Declaration.Enum(... derives0, ...)
```

Reply:
Take and return?

Reply:
Yes, I get it now

Reply:
I think